### ISSUE TITLE
bst should not run autogen or autoreconf for autotools tarballs

### ISSUE THREAD

**Issue by Cynical-Optimist** [buildstream-migration/bst-staging#256]
[See original issue on GitLab](https://gitlab.com/BuildStream/buildstream/-/issues/256)
In GitLab by [[Gitlab user @TheRealMichaelCatanzaro]](https://gitlab.com/TheRealMichaelCatanzaro) on Feb 15, 2018, 17:44

autotools.yaml contains the following code:

```
  autogen: |
    export NOCONFIGURE=1;
    if [ -e autogen ]; then ./autogen;
    elif [ -e autogen.sh ]; then ./autogen.sh;
    elif [ -e bootstrap ]; then ./bootstrap;
    elif [ -e bootstrap.sh ]; then ./bootstrap.sh;
    else autoreconf -ivf;
    fi
```

This is good for builds from git, but it is broken for builds from tarballs. Running autogen/autoreconf blows away the generated files that are distributed with the tarball, which means we have no way to know whether the recommended workflow `./configure; make; sudo make install` actually works or not.

It caused two spurious failures for the GNOME 3.27.90 release, both of which were partially BuildStream's fault:

 * gnome-documents's autogen.sh performs a git submodule update, which immediately fails when run outside of a git repo. autogen.sh really should not be in the dist at all, but it shouldn't matter, because BuildStream should not be executing it in the first place.
 * gedit experienced a [much more subtle build failure](https://bugzilla.gnome.org/show_bug.cgi?id=793459) caused by BuildStream running autoreconf instead of gedit's autogen.sh script. If a module contains autogen.sh, then only autogen.sh should run autoreconf: build failures are to be expected if it is run automatically. gedit properly does not dist autogen.sh, because autogen.sh is only expected to be run from git, so BuildStream went ahead and ran autoreconf anyway, which broke the build.

The fix is for BuildStream to skip running autogen/bootstrap/autoreconf if configure already exists.

Lastly, note that GNOME could care less whether autoreconf runs successfully on the tarball. I know Debian has started to build all its packages that way, to make sure they build cleanly with the newest autofoo. But that's not interesting for GNOME. We want to know that the tarballs build together *as released by GNOME*, generated with whatever different versions of the autotools that happened to be installed on the maintainers' computers.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Feb 15, 2018, 18:01

I'm not convinced of this one.

Should a tarball not behave correctly if built without the expectation of being a tarball ?

Should we have to maintain separate build instructions based on whether we are buidling from git or from a tarball ?

It's first of all probable that there is no submode definitions defi ed for these submodules in the metadata, have we tried defining the metadata to ensure the submodules are checked out at the correct revision before trying to auto-do-it where networking is disabled ?

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @leio]](https://gitlab.com/leio) on Feb 15, 2018, 18:06

If you are running autoreconf, then you really need to do more, like I linked on the GNOME bug. Something like https://gitweb.gentoo.org/repo/gentoo.git/tree/eclass/autotools.eclass#n164

For GNOME, they want to test the tarballs as released, so it makes sense to not regenerate anything and just test what is released. So doing any of that (autoreconf/intltoolize/etc) means that the tarball release of individual releases isn't tested as released, as well, you are regenerating it..

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @jjardon]](https://gitlab.com/jjardon) on Feb 15, 2018, 18:06

For reference, the behaviour used to be as [[Gitlab user @TheRealMichaelCatanzaro]](https://gitlab.com/TheRealMichaelCatanzaro) is requesting, but it was changed in commit be08caa06d6b8ba780aaaaadc56437f10418f39c in MR !148

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @TheRealMichaelCatanzaro]](https://gitlab.com/TheRealMichaelCatanzaro) on Feb 15, 2018, 18:06

> I'm not convinced of this one.
>
> Should a tarball not behave correctly if built without the expectation of being a tarball ?

Uh... the correct way for a tarball to behave is to work when you run `./configure; make; make install`. That's not what BuildStream is doing.

> Should we have to maintain separate build instructions based on whether we are buidling from git or from a tarball ?

Maybe. I think you could get away with just checking whether configure already exists, though. (If it exists and is executable, no reason to run autogen.sh.)

> It's first of all probable that there is no submode definitions defi ed for these submodules in the metadata, have we tried defining the metadata to ensure the submodules are checked out at the correct revision before trying to auto-do-it where networking is disabled ?

Ah, that's another problem! We don't want to ever update submodules for tarball builds; we should be using the version of the submodule that is bundled in the tarball. How would trying to update submodules work when there is no git repo?

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Feb 15, 2018, 18:07

One issue of skipping `autoreconf` when `configure` already exists is that the build may break if `configure.ac` or `Makefile.am` files are patched (depending on autotools configuration).

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @TheRealMichaelCatanzaro]](https://gitlab.com/TheRealMichaelCatanzaro) on Feb 15, 2018, 18:08

> For reference, the bahaviour used to be as [[Gitlab user @TheRealMichaelCatanzaro]](https://gitlab.com/TheRealMichaelCatanzaro) is requesting, but it was changed in commit be08caa06d6b8ba780aaaaadc56437f10418f39c

Almost. To be robust against modules that (either accidentally or intentionally) choose to dist autogen.sh, my request is to not run autogen.sh even if it exists.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @TheRealMichaelCatanzaro]](https://gitlab.com/TheRealMichaelCatanzaro) on Feb 15, 2018, 18:11

> One issue of skipping `autoreconf` when `configure` already exists is that the build may break if `configure.ac` or `Makefile.am` files are patched (depending on autotools configuration).

Good point, if there is a patch to configure.ac or Makefile.am, then it autogen necessarily has to be run, that is true... so some sort of override would be needed.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @TheRealMichaelCatanzaro]](https://gitlab.com/TheRealMichaelCatanzaro) on Feb 15, 2018, 18:22

> Uh... the correct way for a tarball to behave is to work when you run `./configure; make; make install`. That's not what BuildStream is doing.

Keep in mind that the tarballs are supposed to build even when none of the Autotools are installed.

> Good point, if there is a patch to configure.ac or Makefile.am, then it autogen necessarily has to be run, that is true... so some sort of override would be needed.

This is what we do in Fedora, anyway: we just add a call to autoreconf when we need to patch configure.ac or Makefile.am for a given package. It's not expected to work in general, but we can verify that it works for particular packages on an as-needed basis.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @TheRealMichaelCatanzaro]](https://gitlab.com/TheRealMichaelCatanzaro) on Mar 2, 2018, 21:04

This is going to be a problem again for next week's GNOME release.

What if we change autotools.yaml to run autoreconf only if configure is not already present? Then it would not need to know whether it was doing a tarball build or not. What might that break?

If that doesn't work, I think we will need separate logic for tarball builds and git builds. Like it or not, expectations for autotools tarball builds are completely different....

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Mar 2, 2018, 21:21

I also don't like it that the current defaults break in the traditional build scenario: tarball and no autotools installed. Not sure what the best defaults are for BuildStream. I tend towards running autogen only if configure does not exist (as per [Colin's Build API](https://people.gnome.org/~walters/build-api/build-api.md)) but providing a convenient override for the case of patched tarballs.

Independent of whether or how we change the defaults in BuildStream, it's already simple to override the defaults in `project.conf`:
```
elements:
  autotools:
    variables:
      autogen: |
        export NOCONFIGURE=1
        if [ -x configure ]; then true
        elif [ -x autogen ]; then ./autogen
        elif [ -x autogen.sh ]; then ./autogen.sh
        elif [ -x bootstrap ]; then ./bootstrap
        elif [ -x bootstrap.sh ]; then ./bootstrap.sh
        else autoreconf -ivf
        fi
```

I.e., this should be easy to work around for GNOME.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Mar 3, 2018, 09:38

Lets start by applying the above override for gnome-build-meta, changing the defaults is a much larger discussion.

As a next step (or even immediately, *instead* of the above mentioned override), I want to change GNOME's `convert-to-tarballs.py` script to just stop excluding `m4-common` from the conversion; for reasons outlined in the long post I have written below.

As [[Gitlab user @jjardon]](https://gitlab.com/jjardon) mentions, [[Gitlab user @bochecha]](https://gitlab.com/bochecha) fixed the inconsistent behavior in !148, the [email discussion](https://mail.gnome.org/archives/buildstream-list/2017-November/msg00015.html) is worth a read, too.

The original script which was inherited from baserock did something more weird, as it would try to regenerate `./configure` with `./autogen.sh` or `./bootstrap` but then only fallback to `autoreconf -ivf` in the case that `./configure` did not exist; which is really weird.


There are some opinions I'll throw on the table here for discussion.


## Consistency vs convenience

I feel it is clear, and the process is more predictable, if we don't conditionally do things based on what debris is found in the checkout.

If we were to only generate configure if it doesn't exist; then I think it would be more convenient (things might *"just work"* in some cases that otherwise need project side customization), but then we would have a default that does something less predictable.

In other words, a tarball should in every case *be able to build* the same way as you build from a VCS checkout of the same module, with the exception that you have to provide the necessary extra dependencies such as `.m4` files.

BuildStream's focus is to be deterministic and predictable by default, so I think the current default is the sane one - it introduces the least surprise to unsuspecting users.


## Build reliability / stale tarballs

The main reasons why we normally build a tarball with BuildStream (again this is opinion, but based on other use case scenarios), are:
* When bootstrapping a compiler and runtime, and you just dont have the tooling you need yet to generate `./configure`
* Because that source code is not available in a proper upstream VCS - usually in this case the tarballs are quite old and stale.

In the second case, it is not safe to trust an existing `./configure`, it must be regenerated with the more recent `.m4` files found in the modern OS you are building. One example of this, is that new machine architectures are not recognized by old tarballs (e.g. `aarch64` is not known by old configure scripts).

However, other problems are likely to occur when building tarballs which have pre-encoded `.m4` macros from another site, and the build should be considered overall more reliable when using the `.m4` macros which come from the actual dependencies you are building against.

### Why prefer a VCS ?

Adding this in as a subpoint to the previous.

In the greater scheme of things, when you are building/maintaining an integrated stack, which might be an OS, it is IMO desirable to build everything from the upstream VCS at specific tags.

The reason for this is because of how workflow typically happens with downstream patches. Distros tend to maintain patches along side build instruction metadata like rpm spec files. This is for historical reasons mostly, is gross and disgusting, and should die a slow and painful death.

Today we have better VCS's than we did 20 years ago, and we should be maintaining downstream patches in branches, either upstream or in downstream forks, rather than living with the pain of hand maintaining patch files which go stale over time.


## GNOME should not really need this

Setting aside that `convert-to-tarballs.py` is a very weird thing to do in the first place, as it introduces a completely different build from the one normally used, only at release time: The only reason we are having trouble at release time, I suspect; is because of a special case in the said conversion script which fails to include the `m4-common`.

Excluding `m4-common` at conversion time is something inherited from the original conversion script which dealt with JHBuild.

This itself should be removed, and the result would be:
* The builds would work
* The tarballs would be more reliably built against the `m4-common` macros we are releasing *against*, not the `m4-macros` which happened to be on the computer where the tarball was generated


## An alternative feature

We could add a feature to the `autotools` element specifically for this, too.

As [[Gitlab user @juergbi]](https://gitlab.com/juergbi) pointed out, it is easy to override in a project.conf or at an element.bst level; but it's a little bit verbose to do so.

If it's desirable, it would be easy to add a variable which conditionalizes this behavior. With such a switch, we could turn the suggested override into something like the following in `project.conf`:

```
elements:
  autotools:
    variables:
      preserve-configure-script: True
```

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Mar 3, 2018, 11:24

While I also favor the clean way of building from VCS and regenerating everything from sources, I don't agree with all points.

> if we don't conditionally do things based on what debris is found in the checkout.

Pre-generated files in a proper release tarball can hardly be called **debris**. Building without autogen from a release tarball leads to equally reproducible builds as building with autogen from git.

> In the second case, it is not safe to trust an existing `./configure`, it must be regenerated with the more recent `.m4` files found in the modern OS you are building. One example of this, is that new machine architectures are not recognized by old tarballs (e.g. `aarch64` is not known by old configure scripts).

If the tarball is really old, it might not even be buildable with recent autotools. Since especially autoconf development has slowed down a lot, this has become less of an issue. However, there are still packages that need autoconf 2.13 because they haven't been ported yet to more recent autoconf. And I don't know whether similar issues will arise again with the (forever) upcoming autoconf 2.70 and automake 2.0.

In my experience outdated `.m4` files are rarely an issue, however, outdated `config.guess` and `config.sub` missing recent architectures and operating systems / ABIs can indeed be an issue. Simply copying updated `config.guess` and `config.sub` is sufficient, though, there is no need to run `autoreconf` because of this. As far as I know, Debian updates `config.guess` and `config.sub` with the system copies by default (with `dh`) but doesn't run `autoreconf` by default.

# Following tested procedures

With the recommendation of building from VCS (which I fully agree with), the only elements that are built from tarballs are those where building from VCS checkout is problematic for one reason or another. (And those where no public VCS exists but that's hopefully a rare exception.) Due to this I expect that most tarball elements cannot be successfully built with autogen/autoreconf (for projects following the VCS recommendation), so if this is enforced by default, you have to disable it for almost all tarball elements.

There are also many packages where release tarballs don't contain all files from the VCS, especially `autogen.sh` is often left out. And while `autoreconf` works in common cases, it doesn't work for all projects. See the gedit example mentioned above. There are typically two tested approaches of how to build an autotools project:
* from VCS with `autogen.sh`
* from tarball with `configure`

Maintainers normally don't test building tarballs with `autogen.sh`, as far as I can tell. And `distcheck` also doesn't test this. Projects such as coreutils fetch translations as part of `bootstrap`.

I'm still in favor of skipping `autogen` if `configure` exists as that's following the documented and tested instructions and thus requires the least amount of tweaking. If you want to build everything from scratch, follow the recommendation and build from VCS. Otherwise build from tarball and still have everything working out of the box. However, as it's easy to override in a project, I can live with any upstream defaults.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Mar 3, 2018, 11:43

> There are also many packages where release tarballs don't contain all files from the VCS, especially autogen.sh is often left out. And while autoreconf works in common cases, it doesn't work for all projects. See the gedit example mentioned above.

These are fair points, I wanted to get discussion going because I don't think we should ever take changes to our default behavior lightly.

I was also unaware that `config.sub` and `config.guess` are considered without re-running autotools to generate `configure`.

With either default, I think it may still be desirable to add a switch feature to the `autotools` element, with the opposite default we could call it `force-autoreconf`; it seems nice to have this knowledge in the `autotools` element itself. It is also sane to take the approach where an explicit override is less used than the default.

Perhaps a feature might be premature; we should try full rebuilds of the projects we know about first in order to discover if such a feature is useful enough.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @jbicha]](https://gitlab.com/jbicha) on Mar 3, 2018, 12:16

> As far as I know, Debian updates `config.guess` and `config.sub` with the system copies by default (with `dh`) but doesn't run `autoreconf` by default.

Debian packaging using dh and debhelper compat 10 or higher does run autoreconf and update those 2 autotools files by default as part of the build. (That means it is done for virtually all GNOME-related packages built in Debian now.)

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Mar 3, 2018, 13:16

> Debian packaging using dh and debhelper compat 10 or higher does run autoreconf and update those 2 autotools files by default

Interesting. If this now works as default for Debian, it should probably also work as default for BuildStream. However, based on a quick glance it seems Debian always calls `autoreconf -f -i` and ignores `autogen.sh` and co. (by default). I.e., it doesn't match the current BuildStream procedure either.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @TheRealMichaelCatanzaro]](https://gitlab.com/TheRealMichaelCatanzaro) on Mar 3, 2018, 15:59

[[Gitlab user @tristanvb]](https://gitlab.com/tristanvb), right now I'm kinda thinking "square peg round hole"... your expectations for autotools tarballs are very, very different from the expectations of the autotools.

We don't want m4-common in the release because it's only needed when running autogen.sh. If we expected autogen.sh to be run from the tarball, we would not be running make dist at all and there would not be any generated files in the tarball. Since we do release with generated files in the tarball, we want to make sure they are functional *as released*. Missing support for aarch64, for example, is a big problem that needs to be fixed by the maintainers responsible for releasing the tarballs.

I assume Debian is not running autoreconf for gedit, because we see clearly that cannot ever work. I'm going to suggest that !148 is very clearly incorrect. I doubt calling autoreconf on its own is going to work properly for more than a couple GNOME modules; our modules require calls to aclocal, gtkdocize, gettextize, etc. They also have to be done in a specific sequence. Calling autoreconf without knowing which other programs need to be invoked, in which order, is not really expected to work. Only running autogen.sh is expected to work, but autogen.sh is not expected to be in the tarballs at all. We could tell gedit to dist autogen.sh, but that would be the opposite of the normal recommendation, because tarball builds are expected to use the configure script generated by the maintainer.

Now, it may well be worthwhile for GNOME to give up on release tarballs, stop making them entirely, and tell distributions to build from our git tags instead. I don't have any strong preference on that. But that would be a huge change, and it would be nice to not take that step at the same time that we are switching to BuildStream.

We could also start creating our tarballs using `git archive` rather than `make dist`, to avoid generated files. Then the tarballs would unquestionably be expected to behave in exactly the manner [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) desires. But (a) at this point, it seems easier to just keep porting modules to meson rather than argue about the best way to release autotools tarballs, and (b) we don't control our dependencies. And the switch to the Freedesktop 1.8 SDK is going to bring in a bunch more autotools tarballs, so it's likely this will soon be a problem for more than just gedit and gnome-documents.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @TheRealMichaelCatanzaro]](https://gitlab.com/TheRealMichaelCatanzaro) on Mar 3, 2018, 16:03

> Interesting. If this now works as default for Debian, it should probably also work as default for BuildStream. However, based on a quick glance it seems Debian always calls `autoreconf -f -i` and ignores `autogen.sh` and co. (by default). I.e., it doesn't match the current BuildStream procedure either.

I assume it must have more heuristics than that, to decide whether and how to run aclocal, gtkdocize, gettextize, glib-gettextize, or intltoolize.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @TheRealMichaelCatanzaro]](https://gitlab.com/TheRealMichaelCatanzaro) on Mar 3, 2018, 16:09

> Interesting. If this now works as default for Debian, it should probably also work as default for BuildStream.

Also, while it might work as a default for BuildStream, for GNOME our goal is not to figure out whether we can build the tarballs following our own special recipe with its own special logic. The goal is to figure out whether *our users* will be able to build the tarballs following the normal autotools instructions: `./configure; make; sudo make install`.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @jbicha]](https://gitlab.com/jbicha) on Mar 3, 2018, 16:16

>I assume Debian is not running autoreconf for gedit

We're probably getting off-topic here, but Debian does run `autoreconf -f -i` via [dh_autoreconf](https://buildd.debian.org/status/fetch.php?pkg=gedit&arch=amd64&ver=3.27.90-1&stamp=1518020801) for gedit.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Mar 3, 2018, 16:18

> The goal is to figure out whether _our users_ will be able to build the tarballs following the normal autotools instructions

For GNOME I definitely agree. As long as we create release tarballs that contain pre-generated `configure`, we should test the `configure` we ship without regenerating anything.

(The decision could be to test both but we should definitely test what we ship.)

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Mar 4, 2018, 13:01

[[Gitlab user @TheRealMichaelCatanzaro]](https://gitlab.com/TheRealMichaelCatanzaro) I know it's a big thread, I think you missed my [reply to Jürg](https://gitlab.com/BuildStream/buildstream/issues/256#note_61480001).

For this issue, lets go ahead and change these defaults.

Changing defaults should be taken with a grain of paranoia, as things are likely to break somewhere, so before changing this, I think we should first test this change against a full baserock build, since that build is the most complete one we know of (from bootstrap of runtime all the way to image generation iirc).

I have applied the override suggested by Jürg for the GNOME builds, so we should be able to revert that override after we close this bug, but it should not be a blocker for GNOME builds right now.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Jun 8, 2018, 20:50

mentioned in commit 8f2bf4e6e97f3ff859653c6cf60010ce9a2ebca0

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Jun 8, 2018, 20:52

mentioned in merge request !492

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Jun 8, 2018, 21:26

closed via commit 8f2bf4e6e97f3ff859653c6cf60010ce9a2ebca0

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Jun 8, 2018, 21:26

closed via merge request !492
